File: scripts/play_sequence.py
import numpy as np
import time
import csv
from pathlib import Path
import argparse
from tqdm import tqdm
import sys

# Add src directory to Python path for imports
sys.path.append(str(Path(__file__).parent.parent / 'src'))

# Import configuration
from settings.config import FRAME_WIDTH, FRAME_HEIGHT, TARGET_FPS, VIDEO_DURATION

# Import detector and recognizer
from detection.detector import Detector, BlobOutput
from recognition.recognizer import Recognizer

def parse_arguments():
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(description='Play back sequences from input.txt and log recognition results.')
    parser.add_argument('--input-file', 
                       default='input/input.txt',
                       help='Path to input file containing sequence names (default: input/input.txt)')
    parser.add_argument('--output-dir', 
                       default='output/logs',
                       help='Output directory for log files (default: output/logs)')
    return parser.parse_args()

def read_sequences(input_file):
    """Read sequence names from input file."""
    with open(input_file, 'r') as f:
        sequences = [line.strip().replace(' ', '_') for line in f if line.strip()]
    return sequences

def calculate_expected_duration(num_frames, fps=TARGET_FPS):
    """Calculate expected duration in seconds."""
    return num_frames / fps

def play_sequence(sequence_name, output_dir='output/logs'):
    """Play back sequence and log recognition results."""
    # Build paths
    sequence_path = Path('output/sequence') / f'sequence_{sequence_name}.npy'
    labels_path = Path('output/sequence') / f'sequence_labels_{sequence_name}.npy'
    
    # Load sequence
    print(f"\nProcessing sequence: {sequence_name}")
    print(f"Loading sequence from {sequence_path}...")
    try:
        sequence = np.memmap(sequence_path, dtype='uint8', mode='r', shape=(TARGET_FPS * VIDEO_DURATION, FRAME_HEIGHT, FRAME_WIDTH))
        num_frames = len(sequence)
        print(f"Successfully loaded sequence with shape: {sequence.shape}")
    except Exception as e:
        print(f"Error loading sequence: {e}")
        print("Failed to load sequence file.")
        raise
    
    # Load labels if provided
    labels = None
    if labels_path.exists():
        print(f"Loading labels from {labels_path}...")
        labels = np.load(labels_path)
    
    # Initialize detector and recognizer
    detector = Detector()
    recognizer = Recognizer()
    
    # Calculate expected duration
    duration = calculate_expected_duration(num_frames)
    print(f"\nSequence Info:")
    print(f"Number of frames: {num_frames}")
    print(f"Expected duration: {duration:.2f} seconds")
    print(f"Frame rate: {TARGET_FPS} FPS")
    print(f"Frame shape: {sequence[0].shape}")
    print("\nStarting playback...")
    
    # Create output directory
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)
    
    # Create sequence-specific log file
    log_path = output_dir / f'sequence_log_{sequence_name}.csv'
    with open(log_path, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow([
            'frame_index', 
            'recognized_character', 
            'prefix_sum',
            'detection_time_ms',
            'recognition_time_ms',
            'total_frame_time_ms'
        ])
        
        # Process each frame
        start_time = time.time()
        for i in tqdm(range(num_frames), desc="Processing frames"):
            frame_start_time = time.time()
            frame_np = np.array(sequence[i], copy=True)  # Make a copy to ensure we have a contiguous array
            frame = [[int(pixel) for pixel in row] for row in frame_np]  # Convert to list of lists of ints
            
            # Detect blob and measure time
            detection_start = time.time()
            blob = detector.detect(frame)
            detection_time = (time.time() - detection_start) * 1000  # Convert to ms
            
            # Recognize character and measure time
            recognition_start = time.time()
            if blob.prefix_sum > 0:
                character = recognizer.recognize([{'final_prefix_sum': blob.prefix_sum}])
            else:
                character = '_'  # No blob detected
            recognition_time = (time.time() - recognition_start) * 1000  # Convert to ms
            
            # Calculate total frame processing time
            total_frame_time = (time.time() - frame_start_time) * 1000  # Convert to ms
            
            # Log results
            writer.writerow([
                i, 
                character, 
                blob.prefix_sum,
                round(detection_time, 3),
                round(recognition_time, 3),
                round(total_frame_time, 3)
            ])
            
            # Calculate time to sleep to maintain 32 FPS
            elapsed = time.time() - start_time
            target_time = i / 32.0
            sleep_time = max(0, target_time - elapsed)
            time.sleep(sleep_time)
    
    # Calculate final statistics
    total_duration = time.time() - start_time
    actual_fps = num_frames / total_duration
    
    print(f"\nPlayback complete!")
    print(f"Log saved to: {log_path}")
    print(f"\nPerformance Summary:")
    print(f"Total playback duration: {total_duration:.2f} seconds")
    print(f"Average FPS: {actual_fps:.2f}")
    
    # Post-processing: Extract and print the final recognized string
    print("\nRecognition Results:")
    recognized_chars = []
    with open(log_path, 'r') as f:
        reader = csv.DictReader(f)
        for row in reader:
            char = row['recognized_character']
            if char != '_':
                recognized_chars.append(char)
    
    final_string = ''.join(recognized_chars)
    print(f"Final recognized string: {final_string}")
    print(f"Total recognized characters: {len(recognized_chars)}")

def main():
    # Parse arguments
    args = parse_arguments()
    
    try:
        # Read sequences from input file
        sequences = read_sequences(args.input_file)
        print(f"Found {len(sequences)} sequences to process")
        
        # Process each sequence
        for sequence_name in sequences:
            play_sequence(
                sequence_name=sequence_name,
                output_dir=args.output_dir
            )
            
    except Exception as e:
        print(f"Error: {e}")
        return 1
    
    return 0

if __name__ == "__main__":
    exit(main()) 

File: scripts/play_sequence.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <chrono>
#include <thread>
#include <iomanip>
#include <ctime>
#include <filesystem>
#include "src/fpga_sw/detector.hpp"
#include "src/fpga_sw/recognizer.hpp"

namespace fs = std::filesystem;

// Constants from config
const int TARGET_FPS = 32;
const int VIDEO_DURATION = 10; // seconds

struct Frame {
    std::vector<std::vector<uint8_t>> data;
    
    Frame() : data(FRAME_HEIGHT, std::vector<uint8_t>(FRAME_WIDTH)) {}
};

struct Arguments {
    std::string input_file = "../input/input.txt";
    std::string output_dir = "../output/logs";
};

Arguments parse_arguments(int argc, char* argv[]) {
    Arguments args;
    
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--input-file" && i + 1 < argc) {
            args.input_file = argv[++i];
        } else if (arg == "--output-dir" && i + 1 < argc) {
            args.output_dir = argv[++i];
        }
    }
    
    return args;
}

std::vector<std::string> read_sequences(const std::string& input_file) {
    std::vector<std::string> sequences;
    std::ifstream file(input_file);
    
    if (!file) {
        throw std::runtime_error("Failed to open input file: " + input_file);
    }
    
    std::string line;
    while (std::getline(file, line)) {
        if (!line.empty()) {
            // Replace spaces with underscores
            std::replace(line.begin(), line.end(), ' ', '_');
            sequences.push_back(line);
        }
    }
    
    return sequences;
}

void load_sequence(const std::string& sequence_name, std::vector<Frame>& sequence) {
    std::string filename = "../output/sequence/sequence_" + sequence_name + ".npy";
    std::ifstream file(filename, std::ios::binary);
    
    if (!file) {
        throw std::runtime_error("Failed to open sequence file: " + filename);
    }
    
    // Skip numpy header (we know the format)
    file.seekg(0, std::ios::beg);
    
    // Read data directly into frames
    for (auto& frame : sequence) {
        for (int y = 0; y < FRAME_HEIGHT; y++) {
            for (int x = 0; x < FRAME_WIDTH; x++) {
                uint8_t pixel;
                file.read(reinterpret_cast<char*>(&pixel), sizeof(pixel));
                frame.data[y][x] = pixel;
            }
        }
    }
}

void play_sequence(const std::string& sequence_name, const std::string& output_dir) {
    std::cout << "\nProcessing sequence: " << sequence_name << std::endl;
    
    // Initialize sequence
    std::vector<Frame> sequence(TARGET_FPS * VIDEO_DURATION);
    load_sequence(sequence_name, sequence);
    
    std::cout << "Successfully loaded sequence with " << sequence.size() << " frames" << std::endl;
    std::cout << "Frame shape: (" << FRAME_HEIGHT << ", " << FRAME_WIDTH << ")" << std::endl;
    
    // Print frame statistics for the first frame
    int min_val = 255, max_val = 0;
    for (int y = 0; y < FRAME_HEIGHT; y++) {
        for (int x = 0; x < FRAME_WIDTH; x++) {
            min_val = std::min(min_val, (int)sequence[0].data[y][x]);
            max_val = std::max(max_val, (int)sequence[0].data[y][x]);
        }
    }
    std::cout << "First frame statistics:" << std::endl;
    std::cout << "  Min value: " << min_val << std::endl;
    std::cout << "  Max value: " << max_val << std::endl;
    
    // Initialize detector and recognizer
    Detector detector;
    Recognizer recognizer(true);
    
    // Create output directory
    fs::create_directories(output_dir);
    std::string log_filename = output_dir + "/sequence_log_" + sequence_name + "_cpp.csv";
    
    // Open log file
    std::ofstream log_file(log_filename);
    log_file << "frame_index,recognized_character,prefix_sum,detection_time_ms,recognition_time_ms,total_frame_time_ms\n";
    
    // Process each frame
    auto start_time = std::chrono::high_resolution_clock::now();
    std::string recognized_string;
    
    // Progress bar setup
    const int total_frames = sequence.size();
    const int progress_width = 50;
    
    for (size_t i = 0; i < sequence.size(); i++) {
        // Update progress bar
        int progress = (i * progress_width) / total_frames;
        std::cout << "\r[";
        for (int j = 0; j < progress_width; j++) {
            std::cout << (j < progress ? "=" : " ");
        }
        std::cout << "] " << (i * 100) / total_frames << "%";
        std::cout.flush();
        
        auto frame_start_time = std::chrono::high_resolution_clock::now();
        
        // Detect blob
        auto detection_start = std::chrono::high_resolution_clock::now();
        BlobOutput blob = detector.detect(sequence[i].data);
        auto detection_time = std::chrono::duration_cast<std::chrono::microseconds>(
            std::chrono::high_resolution_clock::now() - detection_start).count() / 1000.0;
        
        // Recognize character
        auto recognition_start = std::chrono::high_resolution_clock::now();
        std::vector<BlobOutput> blobs;
        if (blob.prefix_sum > 0) {
            blobs.push_back(blob);
        }
        char character = recognizer.recognize(blobs);
        auto recognition_time = std::chrono::duration_cast<std::chrono::microseconds>(
            std::chrono::high_resolution_clock::now() - recognition_start).count() / 1000.0;
        
        // Calculate total frame processing time
        auto total_frame_time = std::chrono::duration_cast<std::chrono::microseconds>(
            std::chrono::high_resolution_clock::now() - frame_start_time).count() / 1000.0;
        
        // Log results
        log_file << i << "," 
                << character << "," 
                << blob.prefix_sum << ","
                << std::fixed << std::setprecision(3) << detection_time << ","
                << recognition_time << ","
                << total_frame_time << "\n";
        
        // Add character to recognized string if not '_'
        if (character != '_') {
            recognized_string += character;
        }
        
        // Calculate time to sleep to maintain target FPS
        auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(
            std::chrono::high_resolution_clock::now() - start_time).count() / 1000000.0;
        double target_time = i / static_cast<double>(TARGET_FPS);
        double sleep_time = std::max(0.0, target_time - elapsed);
        std::this_thread::sleep_for(std::chrono::duration<double>(sleep_time));
    }
    
    std::cout << "\n"; // New line after progress bar
    
    // Calculate final statistics
    auto total_duration = std::chrono::duration_cast<std::chrono::microseconds>(
        std::chrono::high_resolution_clock::now() - start_time).count() / 1000000.0;
    double actual_fps = sequence.size() / total_duration;
    
    std::cout << "\nPlayback complete!" << std::endl;
    std::cout << "Log saved to: " << log_filename << std::endl;
    std::cout << "\nPerformance Summary:" << std::endl;
    std::cout << "Total playback duration: " << std::fixed << std::setprecision(2) << total_duration << " seconds" << std::endl;
    std::cout << "Average FPS: " << std::fixed << std::setprecision(2) << actual_fps << std::endl;
    std::cout << "\nRecognition Results:" << std::endl;
    std::cout << "Final recognized string: " << recognized_string << std::endl;
    std::cout << "Total recognized characters: " << recognized_string.length() << std::endl;
}

int main(int argc, char* argv[]) {
    try {
        // Parse command line arguments
        Arguments args = parse_arguments(argc, argv);
        
        // Read sequences from input file
        std::vector<std::string> sequences = read_sequences(args.input_file);
        std::cout << "Found " << sequences.size() << " sequences to process" << std::endl;
        
        // Process each sequence
        for (const auto& sequence_name : sequences) {
            play_sequence(sequence_name, args.output_dir);
        }
        
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
} 